### ISSUE TITLE
3 monitor setup problem

### ISSUE THREAD

**Issue by crisidev** [koekeishiya/chunkwm#153]
Hello,
I have a 3 monitor setup with a macbook pro. Monitors are identified like this:
```
|------|------|
|      |      |------|
|  2   |  1   |  3   |
|------|------|------|
```
Where 3 is the laptop monitor.

When I try to switch between monitors with
```
chunkc tiling::monitor -f <id>
```
or
```
chunkc tiling::monitor -f prev / next
```
it mostly does not work.

It reliably works only on transition between 2 -> 1.
It never works between 1 -> 2, 1 -> 3
It never works between 3 -> 1, 3 -> 2

Any idea?

**Comment by koekeishiya** [comment]
the monitor focus commands only try to give focus to the window that macOS says is focused there. If no windows are open or the invisible finder desktop window is in focus, the focus command will fail.

that's not to say that there could not be bugs here, but that's the gist behind the implementation.

**Comment by koekeishiya** [comment]
I noticed that we do not properly include floating window in the above search,
as seen here: https://github.com/koekeishiya/chunkwm/blob/master/src/plugins/tiling/controller.cpp#L1842

That's a bug at least.

**Comment by crisidev** [comment]
I have windows focused in the monitor I am trying to switch focus to and they are not floating..

Tonight I will make some other tests to see if I can give you more background.

**Comment by crisidev** [comment]
It looks like I have the biggest problem when I switch to and from the built in monitor, but the issue is pretty random and I am not sure how to explain it properly.

**Comment by crisidev** [comment]
I am also having focus problem with 3 monitors. Example:
```
|------|------|
|      |      |------|
|  2   |  1   |  3   |
|------|------|------|
```
2: Iterms bsp
1: Chrome
3: Iterm monocle

If I move the mouse from 3 to 2, it always tries to move the focus back to 3.

Again, I have more problems when moving from the integrated monitor to another one.

The switch monitor issue is still there and I haven't found a way to mitigate it.

**Comment by koekeishiya** [comment]
will take a look at this when i have time.

**Comment by crisidev** [comment]
Some updates on the focus problem. It definitely mitigated if I disable the border plugin. It still happen some times, but way more usable than with border plugin active.

I am thinking about some problems in both ffm and border plugins.

**Comment by koekeishiya** [comment]
Does this happen when using 2 monitors as well, or just 3+ ?

**Comment by crisidev** [comment]
I tested it once and it was happening also with 2 monitors. I will retry during the day and keep you posted.

**Comment by crisidev** [comment]
The focus issue and also the monitor switch one happens also with 2 monitors. 

The focus one is happening less often than with 3 monitors. 

The monitor switch still not work from external to built and works sometimes from built in to external.

**Comment by koekeishiya** [comment]
Are you running any apps that creates overlays or similar? Can yiu reproduce with simply having 2monitors with a terminal and browser only?

**Comment by crisidev** [comment]
I think I was able to tackle down the problem. I don't have a clue why it is happening, but at least now I know when. Please note it happens both with and without ubersitch active.

The issue happens only in a specific display / apps configuration, which is the one I use more often when I am working:
```
|------|------|
|      |      |------|
|  2   |  1   |  3   |
|------|------|------|
```

Monitor 2: Iterm
Monitor 1: Browser, both chrome and firefox (actually any application which is not iterm)
Monitor 3: iterm

If I move the mouse between 2 and 3 and vice versa, the focus often go back to the original monitor.

If I have only iterms on all monitors it does not happen. If I change window positioning like
Monitor 2: browser
Monitor 1: iterm
Monitor 3: iterm
it does not happen and also the border plugin does not have problems. 

Here is an extract of chunkwm log when I move from 2 to 3 with a non iterm window in 1:
```
# iterm 136060 is on monitor 3
# iterm 163888 is on monitor 2
# chrome is on monitor 1

1:desktop   163009:0:invalid window:Google Chrome:
1:desktop   163010:0:Google Chrome:3 monitor setup problem · Issue #153 · koekeishiya/chunkwm
33752:iTerm2 activated
84820:Google Chrome deactivated        # move mouse from chrome to iterm (1 -> 2)
iTerm2:Shell:163888 window focused    # iterm focused (2)
iTerm2:Shell:163888 window focused    # again?? (2)
iTerm2:Shell:163060 window focused    # focus go back to iterm on builtin display (3)
5:desktop   163888:0:iTerm2:Shell         # (2)
iTerm2:Shell:163888 window focused    # focus go to right iterm (2)
iTerm2:Shell:163060 window focused    # focus back to builtin display (3)
5:desktop   163888:0:iTerm2:Shell         # (2)
5:desktop   163888:0:iTerm2:Shell         # final automatic switch (2)
iTerm2:Shell:163888 window focused    # everything works now
5:desktop   163888:0:iTerm2:Shell
```

This happens in matter of 300ms.

**Comment by koekeishiya** [comment]
> I don't have a clue why it is happening, but at least now I know when. Please note it happens both with and without ubersitch active.

I really have no idea why, the `window focused` prints come from the callback for the event received from macOS: [here](https://github.com/koekeishiya/chunkwm/blob/master/src/core/callback.cpp#L428)
 
I know one thing for sure. I'm really starting to hate macOS..

This is the code of `FocusMonitor` that actually does anything useful except for parsing input and error correction / cleanup.

```
   # gell windows on destination monitors (bug: ignores floating window as of now)
    WindowIds = GetAllVisibleWindowsForSpace(DestinationSpace);
    if(WindowIds.empty())
    {
        goto dest_space_free;
    }

    # list is sorted by recently foused, index zero is the window that macOS says is focused on this space
    Window = GetWindowByID(WindowIds[0]);

    # make this the focused window for its application
    AXLibSetFocusedWindow(Window->Ref);
    # make this application appear as foreground, if it is not already
    AXLibSetFocusedApplication(Window->Owner->PSN);
```

**Comment by dominiklohmann** [comment]
@crisidev can you replicate the issue using any application other than iTerm2 (e.g. Terminal)?

**Comment by crisidev** [comment]
@dominiklohmann yes, I can replicate it also with Terminal.

**Comment by crisidev** [comment]
And if I move windows around, to avoid having one application between 2 terminals, it is even worse.. I cannot focus anymore the terminals..

It is not happening with 2 monitors, only with 3.

**Comment by crisidev** [comment]
I tested with various applications and if I have app type 1 on central monitor and app type 2 on left and right monitor, the issue is there.

**Comment by koekeishiya** [comment]
I only have one external monitor available, so I can't really work on this issue.

**Comment by crisidev** [comment]
@koekeishiya fair enough. Can you give me some hints (code references) where to start looking at? I can try to debug this on my own, but I need to know where to start:
* Where is the focus window handled? 
* Is it handled differently when there are more than one monitor involved?
* Should this be a problem related only to tiling plugin or also border / ffm ones?

Thanks a lot!

**Comment by koekeishiya** [comment]
The border and ffm plugin should not be related.

Focus window event is handled in `plugin.mm`:`WindowFocusedHandler(..)`
The code to focus the window on a given monitor monitor is `controller.cpp`:`FocusMonitor(..)`.

**Comment by crisidev** [comment]
I gave this a shot, but I don't have enough background nor time to figure out where the problem could be :(

Also, am I the only one with this particular issue?

**Comment by net** [comment]
Similar/same issues here with 1 laptop and 1 external (two screens total).

**Comment by crisidev** [comment]
After some testing I think I figured out something: most of problems comes from display connected via displayport. I have one HDMI port on my macbook and If I connect one of the display over HDMI, it behaves better than the other.

@koekeishiya is this ringing you any bell?

**Comment by koekeishiya** [comment]
@crisidev 

None if this makes sense to me at all. The number of connected displays, how they are connected etc should all be irrelevant.

All we do is receive *window_focused* and *application_activated* events from the accessibility API.
These respond the same way in this case,as it would if you manually focus a window, cmd-tab, or some other application triggers an automatic focus change.

From what I'm hearing, it seems like inconsistencies within the accessibility API.

**Comment by crisidev** [comment]
@koekeishiya Agree about the general nonsense on this behaviours.. kwm was not behaving like this thou, but as far as I understand we cannot compare chunkwm and kwm since the way they work is quite different..

**Comment by koekeishiya** [comment]
@crisidev 

> kwm was not behaving like this thou, but as far as I understand we cannot compare chunkwm and kwm since the way they work is quite different

The difference is that in *kwm* I opted to control and override the macOS focus as I saw fit. This ended up causing other subtle problems, so in *chunkwm* we let macOS control everything completely and just respond in the way that is most appropriate.

**Comment by net** [comment]
@crisidev I'm using DisplayPort, but I've tried HDMI before and had issues.

**Comment by net** [comment]
The monitor switch commands

```
chunkc tiling::monitor -f 1
chunkc tiling::monitor -f 2
```

cause the focus glitch fairly reliably.

Two other issues:

Splitting a node horizontally ( — ) overlaps the bottom window and top window.

iTerm displays higher (y) than it should.

**Comment by dominiklohmann** [comment]
Regarding that iTerm issue... update your iTerm (assuming you're using beta or nightly). There was a bug a few versions ago that had a wrong minimum window height requirement. Build 3.1.beta.6 works for on my end.

**Comment by net** [comment]
Seems I wasn't on the latest version of chunkwm. Updating fixed issue 1 & 2. My fault.

@dominiklohmann That sounds like my issue. Unfortunately 3.1.beta.6 didn't fix it for me. I ended up patching a cheaty fix directly into my SIMBL chunkwm companion.

**Comment by net** [comment]
Correction, it did not fix issue 2. Seems to be caused by a y disagreement between chunkwm and my SIMBL companion. That's likely for me to fix though.

**Comment by koekeishiya** [comment]
I'm unable to reproduce any of this, using 2 monitors (all I have available).

For what its worth, can you reproduce this by having `chunkc set window_focus_cycle all`
and doing `chunkc tiling::window -f east | west` to shift between monitors.
This uses the exact same codepath, and I would expect it to also elicit the same issue.

**Comment by dominiklohmann** [comment]
`chunkc set window_focus_cycle all` doesn't wrap around `first` and `last` monitor yet, so that removes a feature that's otherwise available for single monitor setups.

**Comment by koekeishiya** [comment]
@dominiklohmann 

> chunkc set window_focus_cycle all doesn't wrap around first and last monitor yet, so that removes a feature that's otherwise available for single monitor setups.

Created issue #174 to track this.

**Comment by koekeishiya** [comment]
@crisidev 

So I can actually reproduce what I assume is this behaviour, using only two monitors, by doing the following:

**monitor 1**: firefox window, iterm2 window
**monitor 2**: firefox window, iterm2 window

make sure the firefox window is the focused window on monitor 1
make sure the iterm2 window is the focused window on monitor 2

have iterm2 (on monitor 2) be the active window, perform `chunkc tiling::monitor -f 1`

**Result**: firefox on monitor 1 is focused by us as expected (I know this because I see the border for a split 1/10th of a second), but almost instantly, the firefox window on monitor 2 is made active.

So now we know what happens, but I still don't know why it happens. I assume the following for now:

The logic behind the focusing code, is that we first make the window the focused window for its application. After doing that, we then tell macOS to bring that application to the foreground.
The reason we do it in this order is so that other windows of the application is not brought above (in front of) other applications windows.

**The problem**:
macOS appears to have some other form of focused-window history that we are not currently aware of? Our code is actually running as expected, but macOS overrides our focus (probably during the step where we make the application frontmost) with its own flagged window for that application.

**Solution?** `rm -rf /`

**Comment by crisidev** [comment]
Make sense (especially rm -rf / :D)

During my tests, I tried to change the order you are speaking about above, without actual success in focusing properly. 

I am currently only with 1 external monitor and the built in to cope with the problem, but I can make more tests if needed.

**Comment by crisidev** [comment]
Another thing I noticed with 3 monitors: if I set the window_focus_cycle to all and I have the usual setup 
Iterm - Chrome - Iterm

I start from left Iterm, move the focus with a shortcut to next window, it goes to Chrome. Next again it goes to the iterm on the right. From now on, the focus cycle between Chrome and right Iterm and never goes back to the left iterm.

**Comment by koekeishiya** [comment]
@crisidev 

Before or after the latest changes to master?

regarding master, 
I have not actually tested the implementation, as I'm currently on a linux machine, but they should work in theory. Maybe just an off-by-one error in the wrapping calculation.

**Comment by crisidev** [comment]
I never tried this particular thing before building master 15 minutes ago.

**Comment by koekeishiya** [comment]
Hmm definitely not an off-by-one error. Something is going on when there is a third monitor involved at least.

**Comment by crisidev** [comment]
If I use only one external monitor, the focus cycles properly.

**Comment by blefevre** [comment]
I'm running into an issue with 2 external monitors + a MBP that I think is related to this.

I'm unable to change focus or move windows from display 2 to 3 or vice versa.
In order to move/focus the other display, I must first focus display 1.

I have a similar setup to chrisdev except that my 2 is the MBP, and 1 and 3 are external displays connected by display port.

Here's some logs: https://pastebin.com/6uknVyD7
You can see if I attempt to focus 2 -> 3 or 3 -> 2 it reports that it's the same monitor.

Everything works correctly with just 1 external monitor.

**Comment by crisidev** [comment]
This looks like exactly what is happening to me.

**Comment by szymonkaliski** [comment]
I didn't have time to test properly, but on my 3-display setup:

![screen shot 2017-08-10 at 18 57 19](https://user-images.githubusercontent.com/1040420/29182084-c44a447a-7dfd-11e7-972d-25157a3e52ce.png)

Moving windows between screens seems completely broken - most of the time they disappear (sic!), I can see them when right-click on dock icon, but they are nowhere to be seen... @koekeishiya is there any helpful debugging that I could do?

**Comment by koekeishiya** [comment]
@szymonkaliski Could be an issue with your vertical-arranged display ?

I'm not sure if we can detect this, and we do a transform when moving a window to a different monitor, so that it retains its original position and size for the new monitor: https://github.com/koekeishiya/chunkwm/blob/master/src/plugins/tiling/controller.cpp#L1797

I suppose you could printf the monitor bounds here: https://github.com/koekeishiya/chunkwm/blob/master/src/plugins/tiling/controller.cpp#L1522

**Comment by blefevre** [comment]
It might be worth noting that one of my displays is also vertical.

However changing to landscape and restarting chunkwm didn't resolve the issue.

**Comment by koekeishiya** [comment]
Sorry for not being able to provide much help here, but as I mentioned earlier I only have 1 external monitor, in which everything seems to work fine.

The code used for retrieving information about displays is all in here: https://github.com/koekeishiya/chunkwm/blob/master/src/common/accessibility/display.mm


FocusMonitor: https://github.com/koekeishiya/chunkwm/blob/master/src/plugins/tiling/controller.cpp#L1828

SendWindowToMonitor: https://github.com/koekeishiya/chunkwm/blob/master/src/plugins/tiling/controller.cpp#L1691

**Comment by dominiklohmann** [comment]
Played around with this with a three monitor setup (little bit inconvenient because one monitor is a projector for movies, but figured I could use it for testing).

I can get `chunkc tiling::monitor -f <id>` to work if I set up passthrough binds like this in `khd`:

```
# monitor 1, desktops 1-3
ctrl - 1 -> : chunkc tiling::monitor -f 1
ctrl - 2 -> : chunkc tiling::monitor -f 1
ctrl - 3 -> : chunkc tiling::monitor -f 1

# monitor 2, desktops 4-5
ctrl - 4 -> : chunkc tiling::monitor -f 2
ctrl - 5 -> : chunkc tiling::monitor -f 2

# monitor 3, desktops 6-7
ctrl - 6 -> : chunkc tiling::monitor -f 3
ctrl - 7 -> : chunkc tiling::monitor -f 3
```

`ctrl - <id>` is what I have set up to switch to space `<id>` in system preferences. I am not sure why this works. If there was a way to associate a desktop id with a monitor id this would be a little easier to configure.

If any application has windows on multiple monitors, it seems best to just disable mouse follows focus completely, because it seems to just randomly move the mouse to another monitor.

What I think needs to be happen in `chunkwm` for this to work a little better:
- If a monitor focus change happens because of macOS, do nothing
- If a monitor focus change happens because of focus follows mouse, make sure to re-focus the correct window after setting the active monitor
- If a monitor focus change happens because of `chunkc tiling::window --focus <east|next>`, make sure sure to re-focus the _first_ bsp/monocle tree node after setting the active monitor
- If a monitor focus change happens because of `chunkc tiling::window --focus <west|prev>`, make sure sure to re-focus the _last_ bsp/monocle tree node after setting the active monitor
- When a window gets closed, make sure another window on the same desktop gets focused
  - If there are none, focus Finder (Desktop window) to avoid macOS changing the active monitor
- When changing focus to a monitor without windows on the active desktop, focus Finder (Desktop window)
  - Also make this work with focus follows mouse

**Comment by koekeishiya** [comment]
@dominiklohmann 

I appreciate the information you've gathered, but I have a few inquiries.

> If any application has windows on multiple monitors, it seems best to just disable mouse follows focus completely, because it seems to just randomly move the mouse to another monitor.

This seems to me like deeper issues or an effect or a deeper rooted issue that we probably shouldn't ignore.

Regarding the different ways a monitor can get focused, we have no means with the current structure to track how this change occured (and I'm not sure if this is something that is feasible).

By following your suggestions we would be fighting a lot of the *intended (by Apple)* functionality in macOS, which would lead to a lot of friction, if not done perfectly (see kwm for sample of this).

I suppose the core problem is that mostly all of the design of *chunkwm* has been with the mindset of *let macOS do it things and respond* rather than "*interfering*" for a lack of better words.

> When changing focus to a monitor without windows on the active desktop, focus Finder (Desktop window)

If this works, that seems like a great solution to a difffererent *TODO* that I have. Good idea.
We need to actually start tracking this window though, as it is currently ignored alltogether.
Will probably store this separately from the other window references. 
    
> Also make this work with focus follows mouse

Not so sure how to go about this as the *finder desktop window* is not being considered by the  private macOS API that focus follows mouse uses.

Anyway some solid information. Will think about how we might make some of this achievable in a way that will not cause further problems.

**Comment by dominiklohmann** [comment]
> > If any application has windows on multiple monitors, it seems best to just disable mouse follows focus completely, because it seems to just randomly move the mouse to another monitor.
> 
> This seems to me like deeper issues or an effect of a deeper rooted issue that we probably shouldn't ignore.

I really have no idea what's happening here. It only happens sometimes and I can't even get it to fail in a deterministic way. I will definitely have to play around with this some more, but to me this seems like the main issue at hand.

<hr/>

Regarding the workaround with `chunkc tiling::monitor -f <id>`, can we get a way to associate `desktop id` to `space id`? 

Having these commands be independent of what is configured in mission control would be nice.
```
ctrl - <id> -> : chunkc tiling::monitor -f $(chunkc tiling::query --desktop-to-monitor <id>)
ctrl - left -> : chunkc tiling::monitor -f $(chunkc tiling::query --desktop-to-monitor prev)
ctrl - right -> : chunkc tiling::monitor -f $(chunkc tiling::query --desktop-to-monitor next)
```

**Comment by koekeishiya** [comment]
Can someone build and test the branch `#153-patch` and let me know what happens, thanks.

**Comment by dominiklohmann** [comment]
Tried it, what was it supposed to change? Didn't notice anything.

**Comment by dominiklohmann** [comment]
So I figured out when the focus issue is happening and can reliably reproduce it. You need at least 3 monitors to reproduce the issue, but it makes sense and to fix it chunkwm will have to manually override the window focus given by macOS.

The issue requires 3 monitors and at least 2 applications with windows that can be focused. The monitor order is irrelevant, as is the ordering of the windows.

On monitor `1` focus a window of application `A`, then activate monitor `2` and focus another window of application `A`. Now activate monitor `3` and focus a window of application `B`, If you switch focus to monitor `1` now, macOS will give focus to the last active application on monitor `1`, which happens to be application `A`. macOS will then immediately give focus to monitor `2`, because that is where the most recently used window of application `A` was,, so the focus gets moved to the window of application `A` on monitor `2` instead.

**Comment by koekeishiya** [comment]
Misplaced a line of code, try now.

**Comment by koekeishiya** [comment]
The cause of this issue can be reproduced with a single monitor by utilizing multiple desktops. 

This is actually absurd..

What I'm doing:
```
1. Tell chunkwm to focus firefox windon on desktop 4
 1a. Set focused window and main window of Firefox to window in question
 1b. Make firefox the frontmost process
```

What happens:
```
1a. Both focused and main window property of Firefox changes to the window as we expected.
1b. We call the macOS API to make Firefox the frontmost application, but macOS also performs additional steps:
    1c. macOS activates Firefox and immediately forces the application focus to any window on the current space (as per the Mission Control setting).
    1d. macOS now sets this window as the main and focused window for the application, overriding our previous focus change
    1e. rm -rf /
```

I had more on my mind, but lost my train of thought..Maybe I'll remember later.

**Comment by dominiklohmann** [comment]
The patch fixed it. Looks a bit weird because the wrong monitor gets focused first and then the correct one right after, but it works.

**Comment by koekeishiya** [comment]
Just to be clear; I don't consider this *patch* a fix, but just a means to clarify what is actually going on. Will have to think about what to do from now on and how we can do that without having undeterministic behaviour appearing.

**Comment by dominiklohmann** [comment]
For it to really be a fix it would have to not trigger mouse follows focus twice or preferably not focus the wrong window at all before focusing the correct window.

Do you want me to move the focus override stuff (desktop focus etc.) into another issue?

**Comment by koekeishiya** [comment]
The focus override thing can stay here, as I feel its more or less connected.

**Comment by dominiklohmann** [comment]
Found another issue with the patch: If I already have a window of application `A` on monitor `1`, then creating a new window of application `A` on monitor `2` (through `osascript` or right-clicking the dock icon) will activate monitor `1` before creating the new window.

**Comment by koekeishiya** [comment]
@dominiklohmann 

> When changing focus to a monitor without windows on the active desktop, focus Finder (Desktop window) 

I've tried this, but it does not work giving this window focus..

**Comment by dominiklohmann** [comment]
When I click on the desktop or switch to an empty desktop macOS posts the following notifications (in order):

```
AXApplicationActivated (iff previous app was not Finder)
AXMainWindowChanged
AXFocusedWindowChanged
```

This is no different from focusing a normal window, which is why I assumed this was possible.

**Comment by koekeishiya** [comment]
I'm not actually sure if it is working or not.

I do not receive any focused changed notifications for finder if I use the AX-API to make this finder desktop window focused. I do however get the notifications you are talking about if I manually click the desktop, so I know that we subscribed to the correct window handle. I assume that this is because the actual focusing through the AX API does not work on this window.

The actual issue that I wanted to fix with using this approach was the following:

```
// NOTE(koekeishiya): MacOS does not update focus when we send the window
// to a different desktop using this method. This results in a desync causing
// a bad user-experience.
//
// We retain focus on this space by giving focus to the window with the highest
// priority as reported by MacOS. If there are no windows left on the source space,
// we still experience desync. Not exactly sure what can be done about that.
```

With some minor adjustments to the border plugin, we can fix the above issue by simply focusing the Finder application. As there are no windows opened on the active space, we will by default make Finder focus its desktop window for us. I suspect that in the scenario where a Finder window might be open on the active space of a secondary monitor that Finder will focus that window instead, rendereing this approach useless.

**Comment by dominiklohmann** [comment]
Considering `Accessibility Inspector.app` shows them (Open it, then `Window > Show Notifications` in menu bar) just like any other windows notifications I think this should be possible. 


Found [something on Stackoverflow](https://stackoverflow.com/a/15442985) that might work:
```objc
NSArray *apps = [[NSWorkspace sharedWorkspace] runningApplications];

for (NSRunningApplication *app in apps) {
    if([app.bundleIdentifier.lowercaseString isEqualToString:@"com.apple.finder"]) {
        [app activateWithOptions:NSApplicationActivateAllWindows|NSApplicationActivateIgnoringOtherApps];
        break;
    }
}
```

**Comment by koekeishiya** [comment]
The snippet you posted works only if there are no open finder windows anywhere (on any desktop, on any monitor). If there is a Finder window open *somewhere*, it will activate that window.

The problem with my original approach was slightly less, as it would only focus the Finder window if it was on the *active* desktop of any monitor.

Both approaches fail at actually focusing the desktop window.

**Comment by koekeishiya** [comment]
@dominiklohmann 

> Another question might be how to do what NSApplicationActivateAllWindows|NSApplicationActivateIgnoringOtherApps does on a lower level.

We already use the lower level version for focus atm: https://github.com/koekeishiya/chunkwm/blob/master/src/common/accessibility/element.cpp#L93

**Comment by koekeishiya** [comment]
I recently discovered https://contexts.co 

Briefly looking through their release history it appears they also hit this issue.
This is supposedly fixed (and more) in the newest release, but I have yet to test the app and inspect the kind of accessibility notifications that are triggered in response to what the application does.

**Comment by koekeishiya** [comment]
Interestingly enough, https://manytricks.com/witch/ has the same problem we do (ref: https://github.com/koekeishiya/chunkwm/issues/153#issuecomment-322297903), where as *contexts* does not have that problem. They do not seem to do anything special to achieve this, and they seemingly call *SetFrontProcessWithOptions* as well (so does *Witch*).

**Comment by koekeishiya** [comment]
So I've identified that *Contexts* does not use any private APIs to achieve this behaviour. They do use some of the *CGSSpaces* functions for visual purposes, but they are not necessary for what we are interested in.

There is a function `activateMethodSwitchSpaceAndClickWindow` which handles the case we are interested in. Preventing this function from running makes *Contexts* have the same behaviour as *Witch* and us.

**Comment by crisidev** [comment]
Thanks for the digging. I will be back with my laptop soon and I will be able to help with testing!

**Comment by koekeishiya** [comment]
After doing some more experimentation, the only calls that matter seem to be:
```
CGSMainConnectionID
CGSCopyManagedDisplaySpaces
CGSManagedDisplayGetCurrentSpace
CGSSpaceGetType
CGSCopySpaces
CGSGetActiveSpace
CGSCopySpacesForWindows
CGSAddWindowsToSpaces
CGSRemoveWindowsFromSpaces
```

I have a theory that they use `CGSAddWindowsToSpaces` and `CGSRemoveWindowsFromSpaces` to temporarily add the window to the current space, then give that window focus, remove it from the current space and then force a space transition by activating the applications front window.

I've implemented something like that in chunkwm and it works. The focus order is also preserved for the original space. However, there is a very brief moment where you can actually see the window being added and removed just before the space transition occur. Now this isn't really a problem, but it annoys me, and this does not happen when using *Contexts*

**Comment by koekeishiya** [comment]
*Contexts* is actually unable to handle this problem with multiple monitors as well. This is quite surprising honestly, as the entire application is focused around window switching.

Anyway, I've done some changes;  Please do some testing with https://github.com/koekeishiya/chunkwm/tree/i153 when you have time.

**Comment by koekeishiya** [comment]
I already tested, no progress regarding multi-monitor. It did however fix the single-monitor multi-desktop issue I mentioned above.

**Comment by jessemillar** [comment]
For what it's worth, I'm also experiencing similar issues. In my case, properly cycling focus between windows and sending windows to specific desktops by index depends on the monitor arrangement in System Preferences.

My workstation setup is two external monitors and a MacBook Pro with a Retina display. More details can be given if pertinent.

The following screenshot shows the only System Preferences configuration I've found so far that allows `chunkwm` to function as expected. My MacBook is the center display.

![screen shot 2017-09-14 at 4 41 43 pm](https://user-images.githubusercontent.com/4627837/30457098-c77aa1a6-996b-11e7-914b-fe20a017332e.png)

The above arrangement isn't ideal for me because my desk mandates having my MacBook beneath my two monitors. The below System Preferences arrangement is preferred but causes `chunkwm` to not cycle window focus correctly or send windows to specified desktops as expected. On occasion, attempting to send a window to a specific desktop causes the window in question to disappear (seems to happen more often when sending to desktop `3`).

![screen shot 2017-09-14 at 4 41 12 pm](https://user-images.githubusercontent.com/4627837/30457202-1fe49f40-996c-11e7-93a3-0744598b9d2a.png)

Here are links to my current [`chunkwm` configuration](https://github.com/jessemillar/dotfiles/blob/4704458b0d396afb19cf8bf52b77d4632cd76870/.chunkwmrc) and my [`khd` configuration](https://github.com/jessemillar/dotfiles/blob/4704458b0d396afb19cf8bf52b77d4632cd76870/.khdrc) with the relevant `khd` lines being `25-29` and `81-97`.

**Comment by koekeishiya** [comment]
@jessemillar By desktops, do you mean monitor ? By desktop I'm usually referring to the desktops shown in mission control expose, and their associated number.

If you are experiencing problems with sending a window to another desktop (or monitor), would you open a new separate issue describing that, as I'd like to keep this issue focused on the problem related to focusing windows.

---

This issue has become quite long, so to summarize I'll copy @dominiklohmann's post from earlier:

So I figured out when the focus issue is happening and can reliably reproduce it. You need at least 3 monitors to reproduce the issue, but it makes sense and to fix it chunkwm will have to manually override the window focus given by macOS.

The issue requires 3 monitors and at least 2 applications with windows that can be focused. The monitor order is irrelevant, as is the ordering of the windows.

On monitor 1 focus a window of application A, then activate monitor 2 and focus another window of application A. Now activate monitor 3 and focus a window of application B, If you switch focus to monitor 1 now, macOS will give focus to the last active application on monitor 1, which happens to be application A. macOS will then immediately give focus to monitor 2, because that is where the most recently used window of application A was,, so the focus gets moved to the window of application A on monitor 2 instead.

--

This is tricky to fix, because even if we "try to override" the macOS behaviour by automatically doing another focus command, the focus history will be messed up.

I don't really have any ideas for how to solve this problem.

**Comment by dominiklohmann** [comment]
I have another idea regarding this. What if when focusing another monitor, we activate the most recently used<sup>1</sup> application from the target desktop before activating the monitor? 

Right now, the order is as follows:
- **Status**: active monitor 1 has application A
- **User**: switch to monitor 2
- **chunkwm**: activate monitor 2, whose most recently used window is from application B
- **macOS**: focuses the most recently used window from application B, even if that is on monitor 3

I suggest the following order:
- **Status**: active monitor 1 has application A
- **User**: switch to monitor 2
- **chunkwm**: activate application B that was most recently used on monitor 2
- **chunkwm**: activate monitor 2, whose most recently used window is from application B
- **macOS**: ???

I think this is worth a try. It will require caching the most recently used application per desktop/monitor. 

Thoughts @koekeishiya? Have you tried anything like this before?

<hr />
<sup>1</sup> alternatively: first node iff direction is west/north/prev <b>OR</b> last node iff direction is east/south/next

**Comment by koekeishiya** [comment]
There is no way to *activate a monitor* though. The monitor activation happens by actually giving focus to a window. If we just give focus to an application, that will still give the same behaviour.
The problem in our case is that macOS ignores the *focus window* command that we issue before telling it to focus said application.

The steps are more like this:
 - Status: active monitor 1 has application A
 - User: tell chunkwm to switch to monitor 2
 - chunkwm: which window does macOS is focused for desktop on monitor 2
 - chunkwm: tell application B that it's focused window should now be <window from previous step>
 - chunkwm: activate application B
 - macOS: focuses most recently used window of application B, even if that is on monitor 3

I'm sure if we could activate a monitor before sending the focus comands, this problem would be solved.

We can "fake" activate a monitor by simulating a mouse-click, and I did this in early stages of *kwm*, but we need to be smarter this time and make sure we select a point where there is no underlaying window or ui-element that we click on.

**Comment by dominiklohmann** [comment]
Too bad.

Have you played around with the order in which you focus application and window? (I'm assuming you did)
https://github.com/koekeishiya/chunkwm/blob/60f15ef948c9f7003ac9c013a57934065ad23631/src/plugins/tiling/controller.cpp#L1939-L1940

Additionally, do you have any idea how to focus the desktop of another monitor? It just seems to be one scroll area stretching across all windows, and performing click events or sending AXRaise does nothing on it. If you tell Finder to activate it only activates Finder on the active monitor. Accessibility Inspector only has very sparse information on this. If we were able to do this, we could always just activate the desktop of another monitor before activating the target window.

**Comment by koekeishiya** [comment]
> Have you played around with the order in which you focus application and window? (I'm assuming you did)

This is more or less what I did in the *i153* branch. The problem if we swap these or do it in any other order is that it will first focus the applicaiton, thus focusing the 3rd monitor, and then focus the window we wanted on the 2nd monitor.

> Additionally, do you have any idea how to focus the desktop of another monitor? It just seems to be one scroll area stretching across all windows, and performing click events or sending AXRaise does nothing on it. If you tell Finder to activate it only activates Finder on the active monitor. Accessibility Inspector only has very sparse information on this. If we were able to do this, we could always just activate the desktop of another monitor before activating the target window.

As you said, all of these methods pretty much does nothing. The only way I've found is to actually click an empty region by doing `CGEventCreateMouseEvent` followed by `CGEventPost`.

This is problematic if there is a window at the location where we perform the mouse-click event.
At first I thought we could just probe for an empty position, but that is probably not sufficient, as a user can have the menubar hidden, dock hidden, and a window that takes up the entire desktop area.

**Comment by mickaelperrin** [comment]
As reported by other users, some times when trying to move a window from a monitor to another one it simply disappears.

I was wondering if someone have found a workaround to bring back to front the disappeared window without having to kill / quit the app. That's annoying when you have a process running in the problematic window.

Could it possible that chunkwm implements a "reset" option that brings all windows back to the primary monitor ?

**Comment by jessemillar** [comment]
@mickaelperrin You'll probably want to make a separate issue for that suggestion.

@koekeishiya I've clarified my question above for future reference and created a new issue here: https://github.com/koekeishiya/chunkwm/issues/256.

**Comment by koekeishiya** [comment]
I've identified an issue with how monitor-indexes were resolved, which I believe to be the cause of a lot of the symptoms mentioned here.

With regards to the actual topic of this issue, the focusing of windows, I have yet not found a way that will allow us to do this. Looking through other open source project such as hammerspoon and so on, there does not appear to be any way to work around this problem.

I'll be closing this as a can't fix.
